packages/packagers/js/src/ScopeHoistingPackager.js

```diff
@@ -0,0 +1,1049 @@
+    // then we hoist parcelRequire calls to the top of this asset so side effects run immediately.
+    if (isWrapped && dep && !dep?.meta.shouldWrap && symbol !== false) {
+      let hoisted = this.hoistedRequires.get(dep.id);
+      if (!hoisted) {
+        hoisted = new Map();
+        this.hoistedRequires.set(dep.id, hoisted);
+      }
+
+      hoisted.set(
+        resolvedAsset.id,
+        `var $${publicId} = parcelRequire(${JSON.stringify(publicId)});`,
+      );
+    }
+
+    if (isWrapped) {
+      this.needsPrelude = true;
+    }
+
+    // If this is an ESM default import of a CJS module with a `default` symbol,
+    // and no __esModule flag, we need to resolve to the namespace instead.
+    let isDefaultInterop =
+      exportSymbol === 'default' &&
+      staticExports &&
+      !isWrapped &&
+      dep?.meta.kind === 'Import' &&
+      resolvedAsset.symbols.hasExportSymbol('*') &&
+      resolvedAsset.symbols.hasExportSymbol('default') &&
+      !resolvedAsset.symbols.hasExportSymbol('__esModule');
+
+    // Find the namespace object for the resolved module. If wrapped and this
+    // is an inline require (not top-level), use a parcelRequire call, otherwise
+    // the hoisted variable declared above. Otherwise, if not wrapped, use the
+    // namespace export symbol.
+    let assetId = resolvedAsset.meta.id;
+    invariant(typeof assetId === 'string');
+    let obj =
+      isWrapped && (!dep || dep?.meta.shouldWrap)
+        ? // Wrap in extra parenthesis to not change semantics, e.g.`new (parcelRequire("..."))()`.
+          `(parcelRequire(${JSON.stringify(publicId)}))`
+        : isWrapped && dep
+        ? `$${publicId}`
+        : resolvedAsset.symbols.get('*')?.local || `$${assetId}$exports`;
+
+    if (imported === '*' || exportSymbol === '*' || isDefaultInterop) {
+      // Resolve to the namespace object if requested or this is a CJS default interop reqiure.
+      return obj;
+    } else if (
+      (!staticExports || isWrapped || !symbol) &&
+      resolvedAsset !== parentAsset
+    ) {
+      // If the resolved asset is wrapped or has non-static exports,
+      // we need to use a member access off the namespace object rather
+      // than a direct reference. If importing default from a CJS module,
+      // use a helper to check the __esModule flag at runtime.
+      if (
+        dep?.meta.kind === 'Import' &&
+        exportSymbol === 'default' &&
+        resolvedAsset.symbols.hasExportSymbol('*') &&
+        this.needsDefaultInterop(resolvedAsset)
+      ) {
+        this.usedHelpers.add('$parcel$interopDefault');
+        return `($parcel$interopDefault(${obj}))`;
+      } else {
+        if (IDENTIFIER_RE.test(exportSymbol)) {
+          return `${obj}.${exportSymbol}`;
+        }
+
+        return `${obj}[${JSON.stringify(exportSymbol)}]`;
+      }
+    } else if (!symbol) {
+      // Asset was skipped or not found.
+      // TODO: not sure about this.
```

Comment:
I would say make this `invariant(false)` and we'll see if this ever happens 😄 
At least the tests don't trigger this at the moment.